### PR TITLE
Re-implement NMO: remove CRD, use annotations, add node lease, consume upstream drain library

### PR DESCRIPTION
**Pull request by MoserMichael**
add tests. combine commits for DCO signoff

Signed-off-by: Michael Moser <mmoser@redhat.com>

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Comment by MoserMichael** [comment]
Description on the changes in this pull request:


    the operator is not based on CRD; it is based on node annotations: @beekhof says that's because the openshift people didn't want to add another CRD to the system. To start transition to maintenance mode for a node you have to add the lifecycle.openshift.io/maintenance annotation to the node; where the annotation value is the number of seconds it has to be in maintenance mode. The current state of the show is keps in node annotation: lifecycle.openshift.io/maintenance-status
        the value "active" indicates that you are in maintenance mode.
        the value "ended" means you finished with the maintenance mode.
        lots of other values in between

to end maintenance mode delete the lifecycle.openshift.io/maintenance annotation

    the code now used the upstream drain library

    utilise lease objects; k8s.io/api/coordination/v1beta1/Lease - so that for the transition and duration of the maintenance mode there is an active lease with holderIdentity equal to node-maintenance-operator; that's in addition to cordoning the node and setting the taints. Acquiring the active lease is a prerequisite to transitioning towards maintenance mode; that's what it does first when transitioning towards maintenance mode - it acquires the lease.

    Reconcile is in non blocking/asynchronous manner. Pod eviction may block for up to ten seconds, then it times out and will reconcile call will reschedule in order to resume later. The state transitions do not switch on the value of the maintenance-status annotation, as the system is supposed to adjust and respond to the current state (except for one corner case)

@beekhof will correct me if i got anything wrong in this summary of changes. @beekhof wrote a detailed design of the asynchronous state transitions and i humbly implemented it, (but i still didn't get a bonus this quarter ;-(

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 1]
```diff
@@ -4,33 +4,147 @@ package nodemaintenance
...
+
+	clientset kubernetes.Interface
+	drainer   *drain.Helper
+
+	// used only for the duration of one reconcile call. transient value.
+	// That's ok as there is one go routine that handles reconciles (here)
+	reqLogger      *log.Entry
+	node           *corev1.Node
+	specAnnoData   *ReconcileNodeMaintenanceSpecInfo
+	statusAnnoData *ReconcileNodeMaintenanceStatusInfo
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: Global objects are not an appropriate place to store request specific variables

**Comment by beekhof** [comment]
PRs that break automated tests rarely get reviewed.

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go (hunk 1 above)
Comment (MoserMichael) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: fixed in commit f5e6099201dea9369404157aa62270290a02d5e8

**Comment by beekhof** [comment]
Updated the subject to reflect the work being done in this PR

**Comment by beekhof** [comment]
Each commit should address a single problem and touch as little code as possible to do so. 

In addition to testing changes, "first try: integration test." also includes an unrelated update to "put all reconcile specific fields in special type.".  Please fix.

**Comment by beekhof** [comment]
Also, please remove duplicate sign-off lines.

**Comment by MoserMichael** [comment]
> Each commit should address a single problem and touch as little code as possible to do so.
> 
> In addition to testing changes, "first try: integration test." also includes an unrelated update to "put all reconcile specific fields in special type.". Please fix.

i had a problem in that forgot to sign off one of the commits - the build process doesn't like that. therefore had to merge a bunch of them and force push hem. Sorry, won't happen again.

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 2]
```diff
@@ -55,268 +206,866 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
-	if err != nil {
-		return r.reconcileAndError(instance, err)
+func isLeasePeriodSpecified(lease *coordv1beta1.Lease) bool {
+	return lease.Spec.AcquireTime != nil && lease.Spec.LeaseDurationSeconds != nil
+}
+
+// check if lease is not expired (precondition that both AcquireTime and LeaseDuration are not nil)
+func isLeaseValid(lease *coordv1beta1.Lease, addPaddingToCurrent bool) bool {
+
+	if lease.Spec.AcquireTime == nil {
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: Have you checked how lease objects are being used in recent versions?

File: cmd/evicttestcli/main.go [hunk 3]
```diff
@@ -0,0 +1,302 @@
+package main
```
Comment (yanirq) [inline cmd/evicttestcli/main.go]: where is this file coming from ? is it for manual tests ? why not use it under the relevant package and incorporate that as a unittests or functional tests ?

File: Makefile [hunk 4]
```diff
@@ -89,18 +89,18 @@ manifests: csv-generator
...
-	CLUSTER_NUM_NODES=3 ./cluster/up.sh
+	KUBEVIRT_NUM_NODES=3 ./cluster-up/up.sh
 
 cluster-down:
-	./cluster/down.sh
+	./cluster-up/down.sh
 
 cluster-sync:
-	./cluster/sync.sh
+	./cluster/new-sync.sh
```
Comment (yanirq) [inline Makefile]: i would keep one sync if the other is broken

**Comment by yanirq** [comment]
Can you split the CI part into a different PR ?  It would be easier to review and verify that CI works before this PR.  
Adding newer version for tests is great but there are a lot of additions for laying the grounds for newer versions in the test lanes (also removing some of the old lanes could be done).  
You can then leave only the parts in the e2e tests here that are relevant for the functionality itself after you get the CI PR merged.

**Comment by MoserMichael** [comment]
> Can you split the CI part into a different PR ? It would be easier to review and verify that CI works before this PR.
> Adding newer version for tests is great but there are a lot of additions for laying the grounds for newer versions in the test lanes (also removing some of the old lanes could be done).
> You can then leave only the parts in the e2e tests here that are relevant for the functionality itself after you get the CI PR merged.

The commits have been split in https://github.com/kubevirt/node-maintenance-operator/pull/62 . This PR should be closed.

**Comment by MoserMichael** [comment]
re implemented the same stuff with different sequence of commit in: https://github.com/kubevirt/node-maintenance-operator/pull/62

**Pull request closed without merging**